File: examples/02-intro-cpp/example4.cpp
// =================================================================
//
// File: example2.cpp
// Author(s): Pedro González A01651517, Juan Alcantara A01703947
// Description: This file contains the code to count the number of
//				even numbers within an array. The time this implementation
//				takes will be used as the basis to calculate the
//				improvement obtained with parallel technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <climits>
#include <algorithm>
#include "utils.h"

const int SIZE = 1000000000; //1e9

using namespace std;

// implement your class here
class EvenCalculator {
    private:
        int *array, size, result;
    public:
        EvenCalculator (int *array, int size) : array(array), size(size) {}

        int getResult() const { return result; }

        void calculate() {
            result = 0;
            for (int i = 0; i < size; i++)
                if ( !(array[i] % 2) )
                    result++;
        }
};

int main(int argc, char* argv[]) {
	int *a;
	double ms;

	a = new int[SIZE];
	fill_array(a, SIZE);
	display_array("a", a);

	cout << "Starting..." << endl;
	ms = 0;

	// create object here
    EvenCalculator calc = EvenCalculator(a, SIZE);

    start_timer();
    calc.calculate();
    ms += stop_timer();

	cout << "result = " << calc.getResult() << endl;
	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;

	delete [] a;
	return 0;
}

File: examples/04-java-threads/Example4.java
// =================================================================
//
// File: Example4.java
// Authors:
// Description: This file contains the code to count the number of
//                even numbers within an array using Threads.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

public class Example4 extends Thread {
    private static final int SIZE = 100_000_000;
    private int array[];
    private int result;
    private int start; // ptr array start
    private int finish; // ptr array finish
    
    public Example4(int array[], int start, int finish) {
        this.array = array;
        this.result = 0;
        this.start = start;
        this.finish = finish;
    }

    public int getResult() { return result; }

    public void run() {
        for (int i = start; i < finish; i++) {
            if (array[i]%2 == 0)
                result++;
        }
    }

    public static void main(String args[]) {
        int array[] = new int[SIZE], blocks, res = 0;
        long startTime, stopTime;
        Example4 threads[];
        double acum = 0;

        Utils.fillArray(array);
        Utils.displayArray("array", array);

        // Calculate how many threads we can use
        blocks = SIZE / Utils.MAXTHREADS;
        threads = new Example4[Utils.MAXTHREADS];

        acum = 0;
        System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
        for (int j = 0; j <= Utils.N; j++) {

            // Create threads
            for (int i = 0; i < threads.length; i++) {
                if (i != threads.length-1) {
                    threads[i] = new Example4(array, (i*blocks), ((i+1) * blocks));
                } else {
                    threads[i] = new Example4(array, (i*blocks), SIZE);
                }
            }

            startTime = System.currentTimeMillis();

            // Run threads
            for (int i = 0; i < threads.length; i++) { threads[i].start(); }
            for (int i = 0; i < threads.length; i++) {
                try {
                    threads[i].join();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            stopTime = System.currentTimeMillis();
            acum += (stopTime - startTime);

            // Only get result of last run
            if (j == Utils.N) {
                for (Example4 thread : threads) { res += thread.result; }
            }
        }
        System.out.printf("sum = %d\n", res);
        System.out.printf("avg time = %.5f ms\n", (acum / Utils.N));
    }
}
